enterprise/server/remote_execution/action_merger/action_merger.go

```diff
@@ -117,12 +130,79 @@ func RecordClaimedExecution(ctx context.Context, rdb redis.UniversalClient, exec
 }
 
 // Records a hedged execution in Redis.
-func RecordHedgedExecution(ctx context.Context, rdb redis.UniversalClient, adResource *digest.ResourceName) error {
+func RecordHedgedExecution(ctx context.Context, rdb redis.UniversalClient, adResource *digest.ResourceName, groupIdForMetrics string) error {
 	key, err := redisKeyForPendingExecutionID(ctx, adResource)
 	if err != nil {
 		return err
 	}
-	return rdb.HIncrBy(ctx, key, executionCountKey, 1).Err()
+
+	pipe := rdb.TxPipeline()
+	pipe.HIncrBy(ctx, key, executionCountKey, 1)
+	pipe.HIncrBy(ctx, key, actionCountKey, 1)
+	if _, err = pipe.Exec(ctx); err != nil {
+		return err
+	}
+
+	recordMetrics(ctx, rdb, key, groupIdForMetrics)
+	return nil
+}
+
+// Records a merged execution.
+func RecordMergedExecution(ctx context.Context, rdb redis.UniversalClient, adResource *digest.ResourceName, groupIdForMetrics string) error {
+	key, err := redisKeyForPendingExecutionID(ctx, adResource)
+	if err != nil {
+		return err
+	}
+
+	if err := rdb.HIncrBy(ctx, key, actionCountKey, 1).Err(); err != nil {
+		return err
+	}
+
+	recordMetrics(ctx, rdb, key, groupIdForMetrics)
+	return nil
+}
+
+func recordMetrics(ctx context.Context, rdb redis.UniversalClient, key string, groupIdForMetrics string) {
+	hash := rdb.HGetAll(ctx, key)
+	if err := hash.Err(); err != nil {
+		log.Debugf("Error reading action-merging state from Redis: %s", err)
+	}
+	recordCountMetric(ctx, hash, groupIdForMetrics)
+	recordSubmitTimeOffsetMetric(ctx, hash, groupIdForMetrics)
+}
+
+func recordCountMetric(ctx context.Context, hash *redis.StringStringMapCmd, groupIdForMetrics string) {
+	rawCount, ok := hash.Val()[actionCountKey]
+	if !ok {
+		return
+	}
+	count, err := strconv.Atoi(rawCount)
+	if err != nil {
+		return
+	}
+
+	metrics.RemoteExecutionMergedActionsPerExecution.
+		With(prometheus.Labels{metrics.GroupID: groupIdForMetrics}).
```

Comment:
It seems like this metric is going to be a little inaccurate. We are going to record multiple observations for the same digest every time an action is merged. And I think because RecordMergedExecution and RecordHedgedExecution can both be called it might be observed multiple times for the same execution request.
First request: no observation
Second request with hedge: Observe(1), Observe(2) <-- this seems wrong?
Third request without hedge: Observe(3)
In terms of repeating observations, I think this is the best we can do but we shouldn't count multiples times when a request is hedged.

Reply:
Oh shoot, good catch! Fixed.